src/Serokell/Util/Group.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+-- | This module introduces functions for grouping.
+
+module Serokell.Util.Group
+       ( groupBy
+       , groupSingleBy
+       ) where
+
+import Universum
+
+import Data.List.NonEmpty ((<|))
+
+import qualified Data.HashMap.Strict as HM
+
+{- | Groups elements using results of the given function as keys.
+
+>>> groupBy even [1,2,3,4,5,6]
+fromList [(False,5 :| [3,1]),(True,6 :| [4,2])]
+
+-}
+groupBy :: forall t a b . (Container t, Element t ~ a, Eq b, Hashable b)
+        => (a -> b) -> t -> HashMap b (NonEmpty a)
+groupBy f = foldl' (hmGroup f) mempty
```

Comment:
You don't need to pas `f` function to `hmGroup` because `where` can see `f` from scope